### PR TITLE
wrap: Port to v2 protocol

### PR DESCRIPTION
**Pull request by xclaesse**
Fixes: #8754.

### CODE REVIEW

**Comment by jpakkane** [comment]
Could we do an option flag for this instead? That is, if you do, `meson wrap foo` it will use v1 and `meson wrap --v2 foo` would use version 2? In this way we could merge this faster and can finish the server bits and once everything is done and working delete the v1 code?

**Comment by xclaesse** [comment]
@jpakkane we could, but I don't think it's really needed, the server side is already done. We'll have to keep the v1 server for a while to keep older meson version working.

**Comment by xclaesse** [comment]
The plan I suggest:
1. Rename https://github.com/mesonbuild/wrapdb_v2test to https://github.com/mesonbuild/wrapdb (that repo already exists but is basically empty).
2. Update the 301 redirect rules on the server for the new github URL, that's trivial, I have the access.
3. Merge this PR to make "meson wrap" commands use the new v2 URLs. Older versions of Meson will keep using v1 server. Older .wrap files still will have v1 patch URL that use the same server.
4. (Optional) Update server to use 301 redirect for v1 patch URLs as well.
5. (Optional, after a while, at least a couple meson releases) Decommission the v1 server which will make "meson wrap" commands fail on older meson versions, but patch URL will keep working with simple redirect rules.

**Comment by xclaesse** [comment]
Actually, the only missing piece atm is porting the website https://wrapdb.mesonbuild.com. My plan is to generate static github pages from releases.json in the CI.

**Comment by jpakkane** [comment]
Note that there have been updates to wrapdb since the test repo was created.

**Comment by xclaesse** [comment]
> Note that there have been updates to wrapdb since the test repo was created.

I have a script that import releases, I'll just run it again, not a problem.

**Comment by xclaesse** [comment]
For https://wrapdb.mesonbuild.com my plan is to generate some hotdoc markdown that could be merged directly into the main mesonbuild.com website, but I know nothing about hotdoc... Maybe @MathieuDuponchelle could help.

File: mesonbuild/wrap/wraptool.py [hunk 1]
```diff
@@ -57,38 +58,30 @@ def add_arguments(parser: 'argparse.ArgumentParser') -> None:
...
-        raise SystemExit(data)
-    assert isinstance(jd, dict)
-    return jd
-
-def get_projectlist() -> T.List[str]:
-    jd = get_result(API_ROOT + 'projects')
-    projects = jd['projects']
-    assert isinstance(projects, list)
-    return projects
+def get_releases() -> T.Dict[str, T.Any]:
```
Comment (jpakkane) [inline mesonbuild/wrap/wraptool.py:61]: Eventually this should be cached locally for, say, 1 hour.

**Comment by jpakkane** [comment]
[Something's not quite right there.](https://github.com/mesonbuild/wrapdb_v2test/pull/1/checks?check_run_id=2712001711)

**Comment by xclaesse** [comment]
> [Something's not quite right there.](https://github.com/mesonbuild/wrapdb_v2test/pull/1/checks?check_run_id=2712001711)

I know, was running my import script at the same time, and it screwed things. Long story short, my script cannot import new updates from wrapdb v1 into the existing wrapdb_v2test repo. We have to first decide v1 is RO then run the import script from scratch into a clean new wrapdb repo.

File: mesonbuild/wrap/wraptool.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/wrap/wraptool.py:61]: We could do that, yes. I would like to keep it simple for now and get it merged first, we don't have a cache for v1 neither so it's not a regression anyway.

OOC, is there a way in python std to get platform specific user cache dir?

Comment (jpakkane) [reply mesonbuild/wrap/wraptool.py:61]: [I don't think there is](https://stackoverflow.com/questions/13184414/how-can-i-get-the-path-to-the-appdata-directory-in-python).

**Comment by xclaesse** [comment]
All wraps have been imported into https://github.com/mesonbuild/wrapdb. I think this PR is ready to be merged.

**Comment by xclaesse** [comment]
@jpakkane I added the script that generate the wrapdb table here. That way when we push new releases into wrapdb we can just trigger Meson CI to regenerate the website. We'll probably just have to redirect wrapdb.mesonbuild.com to that page.

**Comment by jpakkane** [comment]
The version list in the  output looks a bit cluttered. Maybe the latest version should be emphasized somehow and separated from the rest?

Screenshot 2021-06-05 013130

**Comment by jpakkane** [comment]
Or maybe only have the latest version on that page? Old versions are probably not that interesting. And the page title and name should be something more descriptive like "WrapDB packages".

**Comment by xclaesse** [comment]
> Or maybe only have the latest version on that page? Old versions are probably not that interesting.

I think older upstream versions could be useful, but we could keep only latest revision for each upstream version?

**Comment by xclaesse** [comment]
What about that

image

It keeps only the latest revision per upstream release, and highlight the first (latest) one.

**Comment by xclaesse** [comment]
TBF, I'm not against keeping only the latest version, anyone has an opinion?

We can always tweak that page later, I mostly wanted to infrastructure in place atm.

**Comment by jpakkane** [comment]
```
tools/regenerate_docs.py:120:1: error: Function is missing a type annotation
tools/regenerate_docs.py:136:24: error: Incompatible types in assignment (expression has type "str", variable has type "List[str]")
tools/regenerate_docs.py:154:5: error: Call to untyped function "generate_wrapdb_table" in typed context
```

**Comment by tp-m** [comment]
I think old versions **are** interesting and we should continue to make them accessible in some way. I can think of multiple reasons, e.g.
1. we don't track API/ABI versions, different versions can have different APIs/ABIs. It will depend on the lib if that's the case or not.
2. the latest version(s) might have certain known bugs/regressions (implementation-wise, or build-wise) that one might want to avoid for now, thus sticking to an older version.

Having said that, I don't think they need to be exposed by default. There could be a global "show older versions" thing, or an expander arrow for each package so that only the latest version is shown and older ones are shown only if the expander is clicked.

**Comment by jpakkane** [comment]
Having the newest one in bold is good enough for me.

**Comment by xclaesse** [comment]
>     2\. the latest version(s) might have certain known bugs/regressions

That's true, and it means we probably should display older revisions too. I tried to make an expander (we can add <details> HTML in the markdown), but not great results. Let's go with just highlight the first in the list, that can be improved later if people have great ideas.

**Comment by xclaesse** [comment]
@jpakkane mypy issues fixed.

**Pull request merged**